src/main/java/picard/illumina/IlluminaBasecallsToFastq.java

```diff
@@ -361,203 +370,279 @@ private void populateWritersFromMultiplexParams() {
+                }
+                if (Defaults.CREATE_MD5) os = new Md5CalculatingOutputStream(os, IOUtil.addExtension(outputPath, ".md5"));
+                return os;
+            } catch (final IOException ioe) {
+                throw new RuntimeIOException("Error opening file: " + outputPath.toUri(), ioe);
+            }
         }
 
         @Override
-        public FastqRecordsForCluster convertClusterToOutputRecord(final ClusterData cluster) {
-            final FastqRecordsForCluster ret = new FastqRecordsForCluster(readStructure.templates.length(), readStructure.sampleBarcodes.length(), readStructure.molecularBarcode.length());
-            final boolean appendTemplateNumberSuffix = ret.templateRecords.length > 1;
-            final boolean appendMolecularBarcodeNumber = ret.molecularBarcodeRecords.length > 1;
+        public void write(final ClusterData rec) {
+            int templateIndex = 0;
+            int sampleBarcodeIndex = 0;
+            int molecularBarcodeIndex = 0;
+
+            for (int i = 0; i < this.numReads; ++i) {
+                final ReadData read = rec.getRead(i);
+                final OutputStream out;
+                final String name;
+
+                switch (read.getReadType()) {
+                    case T:
+                        out = templateOut[templateIndex++];
+                        name = readNameEncoder.generateReadName(rec, appendTemplateNumber ? templateIndex : null);
+                        break;
+                    case B:
+                        out = sampleBarcodeOut[sampleBarcodeIndex++];
+                        name = readNameEncoder.generateReadName(rec, null);
+                        break;
+                    case M:
+                        out = molecularBarcodeOut[molecularBarcodeIndex++];
+                        name = readNameEncoder.generateReadName(rec, appendMolecularBarcodeNumber ? molecularBarcodeIndex : null);
+                        break;
+                    default:
+                        throw new IllegalStateException("Read type other than T/B/M encountered.");
+                }
+
+                writeSingle(out, name, read);
+            }
+        }
 
-            makeFastqRecords(ret.templateRecords, templateIndices, cluster, appendTemplateNumberSuffix);
-            makeFastqRecords(ret.sampleBarcodeRecords, sampleBarcodeIndicies, cluster, false);
-            makeFastqRecords(ret.molecularBarcodeRecords, molecularBarcodeIndicies, cluster, appendMolecularBarcodeNumber);
+        /**
+         * Writes out a single read to a single FASTQ's output stream.
+         */
+        private void writeSingle(final OutputStream out, final String name, final ReadData read) {
+            try {
+                final byte[] bases = read.getBases();
+                final byte[] quals = read.getQualities();
+                final int len = bases.length;
+                for (int i = 0; i < len; ++i) {
+                    quals[i] = (byte) (quals[i] + 33);
```

Comment:
The conditional in SamUtils.phreadToFastq will make this method invalid for vectorization. It might be a negligible performance hit depending on how well JIT optimizes. 

I'll give it a shot and re-benchmark.